### PR TITLE
Origin/refactor/retype functions READY FOR REVIEW

### PR DESCRIPTION
**Pull request by Banshaan**
All functions that utilize http.service have been assigned (hopefully) the right types.
I also added/revised some interfaces so that they could be (better) utilized.

### CODE REVIEW

**Review by philipsens — CHANGES_REQUESTED** [review]
Could you first resolve the git conflicts? 
That way we can have a better representation of what will get merged.

File: src/app/debug/debug.component.ts [hunk 1]
```diff
@@ -17,7 +18,7 @@ export class DebugComponent implements OnInit, AfterViewInit {
   @ViewChild(DebugTreeComponent) debugTreeComponent!: DebugTreeComponent;
   @ViewChild('splitter') splitter: any;
   @ViewChild('bottom') bottom!: ElementRef;
-  currentView: any = {};
+  currentView: CurrentTestView = {} as CurrentTestView;
```
Comment (Matthbo) [inline src/app/debug/debug.component.ts:21]: CurrentTestView has some required properties, now you act like those exist by type asserting with `as CurrentTestView` but this is not the proper way to do this, either make the currentView property also be able to be undefined (`currentView?: CurrentTestView`) or make all properties inside CurrentTestView optional

File: src/app/debug/display/display.component.ts [hunk 2]
```diff
@@ -13,7 +14,7 @@ import { CustomEditorComponent } from '../../custom-editor/custom-editor.compone
 export class DisplayComponent {
   displayReport: boolean = false;
   report: any = {};
-  @Input() currentView: any = {};
+  @Input() currentView: CurrentTestView = {} as CurrentTestView;
```
Comment (Matthbo) [inline src/app/debug/display/display.component.ts:17]: See previous comment about the usage of CurrentTestView

File: src/app/debug/table/table.component.ts [hunk 3]
```diff
@@ -18,11 +26,11 @@ import { TabService } from '../../shared/services/tab.service';
...
   DEFAULT_DISPLAY_AMOUNT: number = 10;
-  metadataCount = 0;
-  viewSettings: any = {
+  metadataCount: number = 0;
+  viewSettings: ViewSettings = {
     defaultView: '',
-    views: [],
-    currentView: {},
+    views: {},
+    currentView: {} as View,
```
Comment (Matthbo) [inline src/app/debug/table/table.component.ts:33]: Same case with this, either make currentView be optional/able to be undefined or give it a proper default value

File: src/app/debug/table/table.component.ts [hunk 4]
```diff
@@ -172,13 +180,13 @@ export class TableComponent implements OnInit, OnDestroy {
...
-  listenForViewUpdate() {
+  listenForViewUpdate(): void {
     this.changeNodeLinkStrategyService.changeNodeLinkStrategy.subscribe(() => {
-      this.httpService.getViews().subscribe((views) => {
+      this.httpService.getViews().subscribe((views: Record<string, View>) => {
         this.viewSettings.views = views;
         this.sortFilterList();
-        let viewToUpdate = Object.keys(this.viewSettings.views).find(
-          (view) => view === this.viewSettings.currentView.name,
+        let viewToUpdate: string | undefined = Object.keys(this.viewSettings.views).find(
```
Comment (Matthbo) [inline src/app/debug/table/table.component.ts:188]: Tiny change to make it look nicer
```suggestion
        let viewToUpdate?: string = Object.keys(this.viewSettings.views).find(
```

File: src/app/debug/table/table.component.ts [hunk 5]
```diff
@@ -189,15 +197,14 @@ export class TableComponent implements OnInit, OnDestroy {
...
       if (Object.keys(this.viewSettings.currentView).length > 0) {
         this.debugReportService.changeView(this.viewSettings.currentView);
       } else {
         this.viewSettings.views = views;
         this.calculateViewDropDownWidth();
-        this.viewSettings.currentViewName = Object.keys(this.viewSettings.views).find(
-          (view) => this.viewSettings.views[view].defaultView,
-        );
+        this.viewSettings.currentViewName =
+          Object.keys(this.viewSettings.views).find((view: string) => this.viewSettings.views[view].defaultView) || '';
```
Comment (Matthbo) [inline src/app/debug/table/table.component.ts:207]: I prefer the `??` operator to the `||` operator here, using `||` will not work as expect in some case (like with booleans) while `??` does. For more info see [Nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing)

File: src/app/debug/table/table.component.ts [hunk 6]
```diff
@@ -283,14 +290,17 @@ export class TableComponent implements OnInit, OnDestroy {
     }
   }
 
-  getStatusColor(metadata: any): string {
-    let statusName = this.viewSettings.currentView.metadataNames.find((name: string) => {
+  getStatusColor(metadata: MetaData): string {
+    let statusName: string | undefined = this.viewSettings.currentView.metadataNames.find((name: string) => {
```
Comment (Matthbo) [inline src/app/debug/table/table.component.ts:294]:
```suggestion
    let statusName?: string = this.viewSettings.currentView.metadataNames.find((name: string) => {
```

**Review by Matthbo — CHANGES_REQUESTED** [review]
I see a lot of `let` keywords to initialize a variable.
Maybe its best to see if its really needed and change it to `const` when the variables aren't being reassigned, try to do this for every bit of code you've been working on so that hopefully at some point its correct in most places

File: src/app/shared/pipes/table-cell-shortener.pipe.ts [hunk 7]
```diff
@@ -7,10 +7,16 @@ import { Subscription } from 'rxjs';
...
+  transform(value: string | boolean): string {
     if (value == undefined) {
       return value;
     }
+    if (typeof value === 'boolean' && value === true) {
+      return 'true';
+    }
+    if (typeof value === 'boolean' && value === false) {
+      return 'false';
+    }
```
Comment (philipsens) [inline src/app/shared/pipes/table-cell-shortener.pipe.ts:19]: The triple `=` is used to ensure the value and the type is a match. So also checking that the value is a `boolean` is redundant. 

In my head, a lot of bells start ringing when I see a condition asserting a boolean value and then returning a boolean value (albeit a string). This makes the code look rushed and not thought out well.
I'd suggest this to best solution:
```suggestion
    if (typeof value === 'boolean') {
      return value.toString();
    }
```

Because it follows a few simple steps: First, you look at the type that was given, and then you return it with the return type of the method.

The way you did it, it might as well could be done like this:
```suggestion
    if (value === true) {
      return 'true';
    }
    if (value === false) {
      return 'false';
    }
```

**Review by philipsens — CHANGES_REQUESTED** [review]

File: src/app/debug/table/table.component.ts [hunk 8]
```diff
@@ -300,29 +310,31 @@ export class TableComponent implements OnInit, OnDestroy {
...
+        });
       });
-    });
   }
 
   openSelected(): void {
     for (const report of this.tableSettings.reportMetadata) {
       if (report.checked) {
-        this.openReport(report.storageId);
+        this.openReport(String(report.storageId));
```
[the remark this answers is not in the document]
Comment (Matthbo) [reply src/app/debug/table/table.component.ts:337]: The constructer of String actually does convert, not sure if there is any drawback to it

File: src/app/debug/table/table.component.ts (hunk 6 above)
Comment (Banshaan) [reply src/app/debug/table/table.component.ts:304]: Unfortunately I couldn't get this to work, I also couldn't find anything online about making a variable optional.

File: src/app/debug/debug-tree/debug-tree.component.ts [hunk 9]
```diff
@@ -29,6 +31,20 @@ export class DebugTreeComponent implements OnDestroy {
   private _currentView: any;
   private lastReport?: Report;
 
+  @Input() set currentView(value: CurrentTestView) {
+    if (this.loaded && this._currentView !== value) {
+      // TODO: Check if the current reports are part of the view
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/debug/debug-tree/debug-tree.component.ts]: Yeah my bad, the TODO comments i place are for myself so i can more easily look up what still needs to be done. Looks like i forgot to delete this one.

File: src/app/shared/pipes/table-cell-shortener.pipe.ts (hunk 7 above)
Comment (Banshaan) [reply src/app/shared/pipes/table-cell-shortener.pipe.ts]: Yeah you're completely right

File: src/app/test/test.component.ts [hunk 10]
```diff
@@ -12,18 +12,24 @@ import { DeleteModalComponent } from './delete-modal/delete-modal.component';
...
 @Component({
   selector: 'app-test',
   templateUrl: './test.component.html',
   styleUrls: ['./test.component.css'],
 })
 export class TestComponent implements OnInit {
-  reports: any[] = [];
+  reports: TestListItem[] = [];
   reranReports: ReranReport[] = [];
   generatorStatus: string = 'Disabled';
```
Comment (MatthijsSmets) [inline src/app/test/test.component.ts:30]: Could you make this a boolean, seeing as this can only be enabled or disabled. If that string value is also shown in the HTML, please make a pipe for it.

File: src/app/test/test.component.ts [hunk 11]
```diff
@@ -197,8 +203,8 @@ export class TestComponent implements OnInit {
     }
   }
 
-  uploadReport(event: any): void {
-    const file: File = event.target.files[0];
+  uploadReport(event: Event): void {
+    const file: File = (event.target as TargetWithFiles)?.files[0];
```
Comment (MatthijsSmets) [inline src/app/test/test.component.ts:207]: I like your solution, but i think it would be more readable to use: HtmlInputElement as the target, look at [this](https://github.com/frankframework/frankframework/blob/7452e2d449bd91d2ab49b88fe50f903a4c4cf555/console/frontend/src/main/frontend/src/app/components/input-file-upload/input-file-upload.component.ts#L19C1-L20C1) solution for reference

File: src/app/shared/interfaces/metadata.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+  numberOfCheckpoints: string;
+  status: string;
+  storageId: string;
+  storageSize: string;
+
+  COMPONENT: string;
+  'CONVERSATION ID': string;
+  'CORRELATION ID': string;
+  'ENDPOINT NAME': string;
+  'NR OF CHECKPOINTS': string;
```
Comment (MatthijsSmets) [inline src/app/shared/interfaces/metadata.ts:17]: Using this hardcoded all caps string as an attribute name is not done. Please make them camelCase and maybe use a transform function to transform the backend response with all caps: "CONVERSION ID" => conversationId into this model.

File: src/app/shared/interfaces/metadata.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  STATUS: string;
+  TIMESTAMP: string;
+
+  [key: string]: string | boolean;
```
Comment (MatthijsSmets) [inline src/app/shared/interfaces/metadata.ts:21]: I feel like this is not necessary for an object thats retrieved from the backend. This should only be used when you do not know what the object can contain.

File: src/app/shared/pipes/table-cell-shortener.pipe.ts [hunk 14]
```diff
@@ -7,10 +7,13 @@ import { Subscription } from 'rxjs';
   standalone: true,
 })
 export class TableCellShortenerPipe implements PipeTransform {
-  transform(value: string): string {
+  transform(value: string | boolean): string {
```
Comment (MatthijsSmets) [inline src/app/shared/pipes/table-cell-shortener.pipe.ts:10]: I'm not sure why this can be a boolean, as far as i know a table cell never has a boolean value.

File: src/app/debug/debug-tree/debug-tree.component.ts [hunk 15]
```diff
@@ -47,30 +61,22 @@ export class DebugTreeComponent implements OnDestroy {
     this.showMultipleAtATimeSubscription.unsubscribe();
   }
 
-  @Input() set currentView(value: any) {
-    if (this._currentView !== value) {
-      // TODO: Check if the current reports are part of the view
```
Comment (MatthijsSmets) [inline src/app/debug/debug-tree/debug-tree.component.ts:52]: Has this TODO been solved or an issue created? If so, please link the issue in this pull request.

File: src/app/debug/display/display.component.ts [hunk 16]
```diff
@@ -13,14 +15,21 @@ import { CustomEditorComponent } from '../../custom-editor/custom-editor.compone
...
   metadataTableVisible: boolean = false;
   @ViewChild('container') container!: ElementRef;
   @Input() containerHeight!: number;
+  @Input() currentView: View = {
+    metadataLabels: [],
+    defaultView: false,
+    crudStorage: false,
+    metadataNames: [],
+    nodeLinkStrategy: '',
+    storageName: '',
```
Comment (MatthijsSmets) [inline src/app/debug/display/display.component.ts:31]: You're setting these attributes to an empty array or string. Is this necessary or should the model maybe changed so that the attributes can be undefined? I've seen you do this in other places as well.

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]
Please take a look at my comments. In some cases it feels like you're solving as you go, instead of taking a step back and trying to work out a better solution.

File: src/app/debug/table/table.component.ts [hunk 17]
```diff
@@ -30,7 +30,17 @@ export class TableComponent implements OnInit, OnDestroy {
   viewSettings: ViewSettings = {
     defaultView: '',
     views: {},
-    currentView: {} as View,
+    currentView: {
+      crudStorage: true,
+      defaultView: true,
+      // eslint-disable-next-line prettier/prettier
```
Comment (philipsens) [inline src/app/debug/table/table.component.ts:36]: Why do you ignore?

File: src/app/test/test.component.ts (hunk 10 above)
Comment (Banshaan) [reply src/app/test/test.component.ts:30]: I could definitely make this a boolean. It's only used in 1 html file (test.component.html), Would a pipe be better than something like making a function that returns the disabled/enabled string or using a ternary in the html?

File: src/app/test/test.component.ts (hunk 11 above)
Comment (Banshaan) [reply src/app/test/test.component.ts:207]: It's not entirely clear to me. Do you mean to change the Event type to HTMLInputElement, or the fyle tpe, or both/something else?

File: src/app/debug/display/display.component.ts (hunk 16 above)
Comment (Banshaan) [reply src/app/debug/display/display.component.ts:31]: The reason i did it this way is because if i make them able to be undefined in the interface, there'll be many errors complaining that the value could be undefined. Should i fix this by having those places first check if they have a value or is there a better way of doing this?

Comment (MatthijsSmets) [reply src/app/debug/display/display.component.ts:31]: Yes it's better to check whether the value is (not) undefined than to set the value to an empty string or array.

File: src/app/test/test.component.ts (hunk 11 above)
Comment (MatthijsSmets) [reply src/app/test/test.component.ts:207]:
```suggestion
    const file: File = (event.target as HtmlInputElement)?.files[0];
```

File: src/app/debug/display/display.component.ts (hunk 16 above)
Comment (Banshaan) [reply src/app/debug/display/display.component.ts]: This has been fixed, instead of setting the properties to empty arrays/strings it now checks whether they have a value or not.

File: src/app/debug/debug-tree/debug-tree.component.ts (hunk 15 above)
Comment (Banshaan) [reply src/app/debug/debug-tree/debug-tree.component.ts:52]: This has been fixed, I used TODOs so i could personally keep track of what i still had to do and this was one i forgot to get rid of it seems

File: src/app/shared/pipes/table-cell-shortener.pipe.ts (hunk 14 above)
Comment (Banshaan) [reply src/app/shared/pipes/table-cell-shortener.pipe.ts]: This issue has been fixed, it had to do with the array it looped over also having a boolean value (checked). I changed the interface around a bit so that this wouldnt cause issues anymore

File: src/app/shared/interfaces/metadata.ts (hunk 13 above)
Comment (Banshaan) [reply src/app/shared/interfaces/metadata.ts]: This has been taken out.

File: src/app/shared/interfaces/metadata.ts (hunk 12 above)
Comment (Banshaan) [reply src/app/shared/interfaces/metadata.ts]: This has been fixed

File: src/app/test/test.component.ts (hunk 11 above)
Comment (Banshaan) [reply src/app/test/test.component.ts]: This has been implemented

File: src/app/test/test.component.ts (hunk 10 above)
Comment (Banshaan) [reply src/app/test/test.component.ts]: This has been implemented

File: src/app/report/edit-display/edit-display.component.ts [hunk 18]
```diff
@@ -66,7 +68,7 @@ export class EditDisplayComponent {
   }
 
   downloadReport(exportBinary: boolean, exportXML: boolean): void {
-    let queryString: string = this.report.xml ? this.report.storageId.toString() : this.report.uid.split('#')[0];
+    const queryString: string = this.report.xml ? this.report.storageId!.toString() : this.report.uid!.split('#')[0];
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/report/edit-display/edit-display.component.ts]: This issue has been fixed.

File: src/app/report/edit-display/edit-display.component.ts [hunk 19]
```diff
@@ -52,9 +54,9 @@ export class EditDisplayComponent {
   }
 
   rerunReport(): void {
-    let reportId: string = this.report.storageId;
-    this.httpService.runDisplayReport(reportId, this.currentView.storageName).subscribe((response) => {
-      let element = document.querySelector('#showRerunResult')!;
+    const reportId: string = String(this.report.storageId);
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/report/edit-display/edit-display.component.ts]: StorageId is now saved as a string in the model.

File: src/app/debug/display/display.component.ts [hunk 20]
```diff
@@ -50,14 +52,18 @@ export class DisplayComponent {
...
+    }
     this.httpService.copyReport(data, 'Test').subscribe(); // TODO: storage is hardcoded, fix issue #196 for this
   }
 
   downloadReport(exportBinary: boolean, exportXML: boolean): void {
-    let queryString: string = this.report.xml ? this.report.storageId.toString() : this.report.uid.split('#')[0];
-    this.helperService.download('id=' + queryString + '&', this.currentView.storageName, exportBinary, exportXML);
+    const queryString: string = this.report.xml ? this.report.storageId : this.report.uid.split('#')[0];
+    if (this.currentView.storageName) {
+      this.helperService.download('id=' + queryString + '&', this.currentView.storageName, exportBinary, exportXML);
```
[the remark this answers is not in the document]
Comment (Matthbo) [reply src/app/debug/display/display.component.ts:65]: Small nitpick, not required imo

[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/debug/display/display.component.ts]: I added the suggestion. As for the earlier question about the functionality, I checked yesterday before my final commit and also 5 minutes ago and everything seems to be working fine on my end.

File: src/app/debug/table/table.component.html [hunk 21]
```diff
@@ -9,42 +9,42 @@
         <i class="fa fa-download px-2"></i><span class="caret"></span>
       </button>
       <div ngbDropdownMenu aria-labelledby="dropdownDownloadTable">
-        <button ngbDropdownItem (click)="downloadReports(true, true)">XML & Binary</button>
-        <button ngbDropdownItem (click)="downloadReports(false, true)">XML <i class="fa fa-info-circle" title="Only a binary file can be uploaded again"></i></button>
-        <button ngbDropdownItem (click)="downloadReports(true, false)">Binary</button>
+        <button ngbDropdownItem (click)="downloadReports(true, true)"> XML & Binary </button>
+        <button ngbDropdownItem (click)="downloadReports(false, true)"> XML <i class="fa fa-info-circle" title="Only a binary file can be uploaded again"></i></button>
+        <button ngbDropdownItem (click)="downloadReports(true, false)"> Binary </button>
```
Comment (philipsens) [inline src/app/debug/table/table.component.html:14]: Where are the spaces for? :o

**Review by Matthbo — APPROVED** [review]

File: src/app/debug/table/table.component.ts [hunk 22]
```diff
@@ -475,22 +523,22 @@ export class TableComponent implements OnInit, OnDestroy {
...
-    } else {
+    } else if (this.viewSettings.currentView?.storageName) {
       this.helperService.download(queryString, this.viewSettings.currentView.storageName, exportBinary, exportXML);
     }
   }
 
-  uploadReports(event: any): void {
-    const file: File = event.target.files[0];
+  uploadReports(event: Event): void {
+    const file: File = (event.target as TargetWithFiles)?.files[0];
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:532]: Please use HtmlInputElement like I've suggested before [here](https://github.com/wearefrank/ladybug-frontend/pull/391#discussion_r1584596678)

File: src/app/report/edit-display/edit-display.component.ts [hunk 23]
```diff
@@ -174,19 +176,21 @@ export class EditDisplayComponent {
     let checkpointId: string = '';
     let storageId: string;
     if (this.report.xml) {
-      storageId = this.report.storageId;
+      storageId = this.report.storageId ?? '';
     } else {
-      storageId = this.report.uid.split('#')[0];
-      checkpointId = this.report.uid.split('#')[1];
+      storageId = this.report.uid!.split('#')[0];
+      checkpointId = this.report.uid!.split('#')[1];
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/edit-display.component.ts:182]:
```suggestion
      if (this.report.uid) {
        const ids = this.report.uid.split('#');
        storageId = ids[0];
        checkpointId = ids[1];
      }
```

File: src/app/shared/interfaces/metadata.ts [hunk 24]
```diff
@@ -0,0 +1,10 @@
+import { DebugListItem } from './debug-list-item';
+
+export interface MetaData extends DebugListItem {
+  component: string;
+  conversationId: string;
+  endpointName: string;
+  timestamp: string;
+
+  // [key: string]: string | boolean;
```
Comment (MatthijsSmets) [inline src/app/shared/interfaces/metadata.ts:9]: please remove commented code

File: src/app/shared/interfaces/target-with-files.ts [hunk 25]
```diff
@@ -0,0 +1,3 @@
+export interface TargetWithFiles extends EventTarget {
```
Comment (MatthijsSmets) [inline src/app/shared/interfaces/target-with-files.ts:1]: Look at my comment about using HtmlInputElement, that way this interface is redundant and can be removed

File: src/app/shared/interfaces/view-settings.ts [hunk 26]
```diff
@@ -0,0 +1,8 @@
+import { View } from './view';
+
+export interface ViewSettings {
+  currentView?: View;
+  currentViewName?: string;
```
Comment (MatthijsSmets) [inline src/app/shared/interfaces/view-settings.ts:5]: `currentView` already has an attribute called `name`

File: src/app/shared/pipes/table-cell-type-checker.pipe.ts [hunk 27]
```diff
@@ -0,0 +1,18 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({
+  name: 'tableCellTypeChecker',
+  standalone: true,
+})
+export class tableCellTypeChecker implements PipeTransform {
```
Comment (MatthijsSmets) [inline src/app/shared/pipes/table-cell-type-checker.pipe.ts:7]:
```suggestion
export class TableCellTypeChecker implements PipeTransform {
```

File: src/app/test/test-folder-tree/test-folder-tree.component.ts [hunk 28]
```diff
@@ -20,7 +20,7 @@ export class TestFolderTreeComponent {
     folderBehaviourOnClick: 'select',
     expandAllFolders: true,
     highlightOpenFolders: false,
-    autoSelectCondition: (item: CreateTreeItem) => this.autoSelectItem(item),
+    autoOpenCondition: (item: CreateTreeItem) => this.autoSelectItem(item),
```
Comment (MatthijsSmets) [inline src/app/test/test-folder-tree/test-folder-tree.component.ts:23]: `autoOpenCondition` is an entirely different function in the tree. Why did you change this?

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]
Please take a look at my [previous comment](https://github.com/wearefrank/ladybug-frontend/pull/391#pullrequestreview-2030795834) about taking a step back and deciding what you need to do instead of solving as you go.
